test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/Helpers/NupkgDownloadTimeoutHttpClientHandler.cs

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace NuGet.Protocol.FuncTest.Helpers
+{
+    /// <summary>
+    /// A NuGet feed where every second nupkg download times out, starting from the first attempt.
+    /// </summary>
+    internal class NupkgDownloadTimeoutHttpClientHandler : MockV3ServerHttpClientHandler
+    {
+        bool _lastDownloadFailed;
+
+        public NupkgDownloadTimeoutHttpClientHandler(IEnumerable<string> packagePaths)
+            : base(packagePaths)
+        {
+            _lastDownloadFailed = false;
+            FailedDownloads = 0;
+        }
+
+        public int FailedDownloads { get; private set; }
+
+        protected override async Task<HttpResponseMessage> GetPackageDownloadResponse(string uri)
+        {
+            HttpResponseMessage response = await base.GetPackageDownloadResponse(uri);
+
+            if (response.StatusCode == System.Net.HttpStatusCode.OK)
+            {
+                _lastDownloadFailed = !_lastDownloadFailed;
+                if (_lastDownloadFailed)
```

Comment:
If I understand correct 1st attempt will result in partial read, but 2nd one is full. Is it right?

Reply:
First attempt will throw an exception, second attempt will not, and then the cycle repeats.